File: exchanges/upbit/candles.js
const { fetchUpbit } = require("./client");

const INTERVAL_PATH = {
  minutes15:  "/candles/minutes/15",
  minutes60:  "/candles/minutes/60",
  minutes240: "/candles/minutes/240",
  days:       "/candles/days",
  weeks:      "/candles/weeks",
  months:     "/candles/months",
};

async function getCandles(market = "KRW-BTC", count = 90) {
  return fetchUpbit(`/candles/days?market=${market}&count=${count}`);
}

async function getCandlesBulk(market = "KRW-BTC", count = 200, interval = "days") {
  const path = INTERVAL_PATH[interval] || "/candles/days";
  count = Math.min(count, 1000);

  const result = [];
  let remaining = count;
  let toParam = "";

  while (remaining > 0) {
    const batchSize = Math.min(remaining, 200);
    const url = toParam
      ? `${path}?market=${market}&count=${batchSize}&to=${toParam}`
      : `${path}?market=${market}&count=${batchSize}`;
    const batch = await fetchUpbit(url);
    if (!batch.length) break;
    result.push(...batch);
    remaining -= batch.length;
    if (batch.length < batchSize) break;
    const oldest = batch[batch.length - 1];
    const oldestDt = new Date(oldest.candle_date_time_utc + "Z");
    oldestDt.setSeconds(oldestDt.getSeconds() - 1);
    toParam = oldestDt.toISOString().replace(".000Z", "");
  }

  return result;
}

module.exports = { getCandles, getCandlesBulk };

File: exchanges/upbit/candles.py
from datetime import datetime, timedelta
from .client import get

INTERVAL_PATH = {
    "minutes15":  "/candles/minutes/15",
    "minutes60":  "/candles/minutes/60",
    "minutes240": "/candles/minutes/240",
    "days":       "/candles/days",
    "weeks":      "/candles/weeks",
    "months":     "/candles/months",
}


def get_candles(market="KRW-BTC", count=90):
    return get("/candles/days", params={"market": market, "count": count})


def get_candles_bulk(market="KRW-BTC", count=200, interval="days"):
    path = INTERVAL_PATH.get(interval, "/candles/days")
    count = min(count, 400)

    batch1 = get(path, params={"market": market, "count": min(count, 200)})
    if len(batch1) < 200 or count <= 200:
        return batch1

    oldest_dt = datetime.strptime(batch1[-1]["candle_date_time_utc"], "%Y-%m-%dT%H:%M:%S")
    to_str = (oldest_dt - timedelta(seconds=1)).strftime("%Y-%m-%dT%H:%M:%S")

    batch2 = get(path, params={
        "market": market,
        "count": min(count - 200, 200),
        "to": to_str,
    })

    return batch1 + batch2
